app/Http/Controllers/Backend/Social/AbstractTwitterController.php

```diff
@@ -0,0 +1,150 @@
+            self::updateToken($socialProfile->user, $socialiteUser);
+            return $socialProfile->user;
+        }
+
+        if (auth()->check()) {
+            self::updateToken(auth()->user(), $socialiteUser);
+            return auth()->user();
+        }
+        return self::createUser($socialiteUser);
+    }
+
+    private static function createUser(SocialiteUser $socialiteUser): User {
+        $user = User::create([
+                                 'name'     => SocialController::getDisplayName($socialiteUser),
+                                 'username' => SocialController::getUniqueUsername($socialiteUser->getNickname()),
+                             ]);
+        self::updateToken($user, $socialiteUser);
+        return $user;
+    }
+
+    private static function updateToken(User $user, SocialiteUser $socialiteUser): void {
+        $user->socialProfile->update([
+                                         'twitter_id'               => $socialiteUser->id,
+                                         'twitter_token'            => $socialiteUser->token,
+                                         'twitter_refresh_token'    => $socialiteUser->refreshToken,
+                                         'twitter_token_expires_at' => Date::now()->add('second', $socialiteUser->expiresIn)
+                                     ]);
+    }
+
+
+    /**
+     * @param Status $status
+     * @param string $socialText
+     *
+     * @return int
+     * @throws NotConnectedException
+     * @throws TweetNotSendException
+     */
+    public abstract function postTweet(Status $status, string $socialText): int;
+
+    /**
+     * @throws NotConnectedException
+     */
+    public static function postStatus(Status $status) {
+        $controller = self::forUser($status->user);
+
+        try {
+            $socialText = self::generateFullSocialText($status);
+            $tweetId    = $controller->postTweet($status, $socialText);
+            $status->update(['tweet_id' => $tweetId]);
+            Log::info("Posted on Twitter: " . $socialText);
+        } catch (NotConnectedException $exception) {
+            throw $exception;
+        } catch (TweetNotSendException $exception) {
+            $status->user->notify(new TwitterNotSent($exception->getStatusCode(), $exception->getStatus()));
+        } catch (Exception $exception) {
+            report($exception);
+            // The Twitter adapter itself won't throw Exceptions, but rather return HTTP codes.
+            // However, we still want to continue if it explodes, thus why not catch exceptions here.
+        }
+    }
+
+    protected static function generateFullSocialText(Status $status): string {
+        $socialText = $status->socialText;
+        if ($status->user->always_dbl) {
+            $socialText .= ' #dbl';
+        }
+        $socialText .= ' ' . url('/status/' . $status->id);
+
+        return $socialText;
+    }
+}
+
+class TweetNotSendException extends Exception
```

Comment:
What is this used for? The `$status` is already known above and could be removed here. Please move to `App\Exceptions`.